feat(react): enable textarea to limit text input to set amount of words

matejoslav opened the pull request:
Co-authored-by: SamChinellato <49278203+SamChinellato@users.noreply.github.com>
Co-authored-by: stevenpatrick009 <106097350+stevenpatrick009@users.noreply.github.com>

Closes #13788 

Added functionality to textArea component to limit the number of words the user can type. Also updated the useAnouncer to handle new wording by allowing it to either display "x characters left" or "x words left" based on args passed in.

#### Changelog

**New**

- textArea component has now word counter mode and can now limit the amount of words user types
- added more tests for the textArea component

**Changed**

- useAnouncer can now take the entityName and use that to display the "x <entity name> left"
- added tests for useAnnouncer



#### Testing / Reviewing

- run the storybook for react components
- navigate to the textArea component
- select the "WithWordCounter" textArea type
- type in few words and observe the counter increase
- try to type in more words than the limit and observe that user can not add more words
- try to paste in text with more words than the limit and observe that the text has been cut to match the word limit

matejoslav committed:
feat(react): enable textarea to limit text input to set amount of words

Co-authored-by: SamChinellato <49278203+SamChinellato@users.noreply.github.com>
Co-authored-by: stevenpatrick009 <106097350+stevenpatrick009@users.noreply.github.com>

matejoslav committed:
Merge branch 'main' into text-area-word-limit

tw15egan committed:
Merge branch 'main' into text-area-word-limit

tw15egan committed:
Merge branch 'main' into text-area-word-limit

matejoslav committed:
test(textArea): fix tests and component

Fix tests affected by the merge to properly use the selected testing library.
Change TextArea component to remove duplicated use of the useRef hook.

matejoslav committed:
docs(developer handbook): change max characters for line in commit body

matejoslav committed:
Merge remote-tracking branch 'upstream/main' into text-area-word-limit

matejoslav committed:
fix(docs): regenerated contributors to resolve merge conflict

matejoslav committed:
Merge branch 'main' into text-area-word-limit

matejoslav committed:
Merge remote-tracking branch 'upstream/main' into text-area-word-limit

tw15egan committed:
Merge branch 'main' into text-area-word-limit

matejoslav committed:
Merge remote-tracking branch 'upstream/main' into text-area-word-limit

alisonjoseph commented on package.json:
```diff
@@ -90,7 +90,7 @@
         2,
         "always",
-        80
+        90
```
What was the reason for this change?

matejoslav replied:
This was because one of the collaborators name was longer than 80 characters and could not be properly included in the commit without increasing the limit for each line. It was the first commit in the branch from December last year

tw15egan replied:
It doesn't have to do with the user's name, it's only validating the character count of the commit message. It just needed a shorter commit message.

matejoslav replied:
Reverted this change

alisonjoseph commented on packages/react/src/components/TextArea/TextArea.js:
```diff
@@ -183,6 +229,11 @@ TextArea.propTypes = {
+  /**
+   * Specify the method used for calculating the counter number
+   */
+  counterMode: PropTypes.oneOf(['default', 'words']),
```
Instead of default and words, would it make more sense to be a little more specific and use `character` and `word`

alisonjoseph reviewed (changes requested):
This is looking and working great! Couple small suggestions, I'm not sure we need to keep the story for this, leaving it in the playground should be sufficient. I would also update the comment / description for the `enableCounter` prop to just say `Specify whether to display the counter` since it is not longer only used for character count.

tw15egan reviewed (approved):
LGTM once the comments from Alison have been addressed 👍🏻 ✅

tw15egan reviewed (changes requested):
It looks like this file was recently converted to Typescript by the Typescript working group helping improve our library. The merge may be a bit messy, it may be best to close this and resubmit starting from the `TextArea.tsx` file

matejoslav committed:
Merge remote-tracking branch 'upstream/main' into text-area-word-limit

matejoslav committed:
revert sync file changes

matejoslav committed:
fix(TextArea): fixed eslint complaints

matejoslav committed:
refactor(TextArea): changes to counterMode and storybook

changed wording for counterMode to be character or word
removed storybook entry

matejoslav committed:
refactor(TextArea): removed unused test

matejoslav committed:
Merge branch 'main' into text-area-word-limit

matejoslav committed:
docs(README): added contributors

matejoslav committed:
Merge branch 'main' into text-area-word-limit

tw15egan commented on tasks/sync.js:
```diff
@@ -9,14 +9,6 @@
-const prettierOptions = {
-  ...packageJson.prettier,
-  parser: 'markdown',
-};
```
Are these changes intentional?

matejoslav replied:
ESLint complained locally that these were unused, will check again

matejoslav replied:
<img width="1126" alt="Screenshot 2023-05-09 at 19 00 19" src="https://github.com/carbon-design-system/carbon/assets/8973672/4cadd0f6-a632-436d-b562-7ff6a2413404">

tw15egan replied:
Okay, I was just unsure why this was being pulled in to this PR. Thanks for clarifying.

alisonjoseph replied:
Agree, this seems unrelated to the issue/PR

matejoslav committed:
revert(package.json & developer-handbook.md): reverted 90 limit to 80

matejoslav committed:
Merge branch 'text-area-word-limit' of github.com:matejoslav/carbon into text-area-word-limit

matejoslav committed:
Merge branch 'main' into text-area-word-limit

matejoslav committed:
Merge branch 'main' into text-area-word-limit

matejoslav committed:
Merge remote-tracking branch 'upstream/main' into text-area-word-limit

tw15egan committed:
chore(format): run yarn format

tw15egan reviewed (dismissed):
This looks great, thanks for sticking with the PR and adding in this functionality! LGTM 👍 ✅

matejoslav committed:
fix(README): fixed the contributors formatting

francinelucca commented on packages/react/src/components/TextArea/TextArea.tsx:
```diff
@@ -131,6 +131,11 @@ export interface TextAreaProps
+  /**
+   * Specify the method used for calculating the counter number
+   */
+  counterMode?: string;
```
```suggestion
  counterMode?: 'character' | 'word';
```

matejoslav replied:
done

francinelucca commented on packages/react/src/components/TextArea/TextArea.tsx:
```diff
@@ -155,22 +161,67 @@ const TextArea = React.forwardRef((props: TextAreaProps, forwardRef) => {
+      ? counterMode === 'character'
+        ? value.toString().length
+        : value.toString().match(/\w+/g)?.length || 0
+      : 0
```
this is really hard to read, could we offload this logic onto a `getInitialTextCount()` function and use if statements instead?

matejoslav replied:
done

francinelucca commented on packages/react/src/components/TextArea/TextArea.tsx:
```diff
@@ -155,22 +161,67 @@ const TextArea = React.forwardRef((props: TextAreaProps, forwardRef) => {
+          setTextCount(evt.target.value.length);
+        } else if (counterMode == 'word') {
+          if (!evt.target.value) {
+            console.log('reset');
```
```suggestion
```

matejoslav replied:
removed

francinelucca commented on packages/react/src/components/TextArea/TextArea.tsx:
```diff
@@ -155,22 +161,67 @@ const TextArea = React.forwardRef((props: TextAreaProps, forwardRef) => {
+              setTextCount(maxCount);
+              textareaRef.current.value = first_max;
+            }
+            // todo: add handling if no matched words were found
```
is a handler still necessary or can we remove this comment?

matejoslav replied:
done

francinelucca commented on packages/react/src/internal/useAnnouncer.js:
```diff
@@ -5,9 +5,12 @@
  */
 
-export function useAnnouncer(textCount, maxCount) {
+export function useAnnouncer(textCount, maxCount, entityName) {
```
if we set the default of `entityName `to 'characters' here then the if statement below is not necessary and we can always construct the sentence using `${entityName}`
```suggestion
export function useAnnouncer(textCount, maxCount, entityName = 'characters') {
```

matejoslav replied:
done

francinelucca reviewed (changes requested):
This is looking great! just some questions and refactoring suggestions

matejoslav committed:
Merge remote-tracking branch 'upstream/main' into text-area-word-limit

matejoslav committed:
Merge remote-tracking branch 'upstream/main' into text-area-word-limit

matejoslav committed:
refactor(TextArea): function to get initial text count

matejoslav committed:
refactor(useAnnouncer): made entityName have default value

francinelucca commented on packages/react/src/components/TextArea/TextArea.tsx:
```diff
@@ -155,31 +161,103 @@ const TextArea = React.forwardRef((props: TextAreaProps, forwardRef) => {
+        ? counterMode === 'character'
+          ? value.toString().length
+          : value.toString().match(/\w+/g)?.length || 0
+        : 0
```
can we call the getinitialtextcount function here as well?

matejoslav replied:
My bad must have missed it

matejoslav committed:
refactor(TextArea): use function to calculate text count in useEffect

francinelucca reviewed (changes requested):
Confirmed no regression on https://github.com/carbon-design-system/carbon/pull/13449 as a result of these changes. 
However, I do see a weird behavior in that in the current state of the TextArea, setting `enableTextCounter` to false makes the component ignore `maxCount` and doesn't look like that is the case for when counter = 'word' with these new changes, see video below: 

https://github.com/carbon-design-system/carbon/assets/40550942/90f33e28-da71-4d85-9f8d-7c49371eabab

I'd expect counter=word and counter=character to behave the same when it comes to `enableCounter` being on/off and the maxCount being reached, which in production is: the `maxCount` is ignored if `enableCounter` is turned off

matejoslav committed:
Merge branch 'main' into text-area-word-limit

matejoslav committed:
Merge branch 'main' into text-area-word-limit

matejoslav committed:
Merge branch 'main' into text-area-word-limit

andreancardona committed:
Merge branch 'main' into text-area-word-limit

matejoslav committed:
Merge branch 'main' into text-area-word-limit

matejoslav committed:
Merge branch 'main' into text-area-word-limit

francinelucca committed:
Merge branch 'main' into text-area-word-limit

matejoslav committed:
Merge remote-tracking branch 'upstream/main' into text-area-word-limit

matejoslav committed:
fix(TextArea): ignore max count limit when counter is disabled

tay1orjones reviewed (dismissed):
I think with that this should be good to go. @francinelucca can you confirm?

matejoslav committed:
Merge remote-tracking branch 'upstream/main' into text-area-word-limit

matejoslav committed:
fix: fixed contributors list

matejoslav committed:
chore(TextArea): trim words on paste and add missing tests

matejoslav committed:
Merge branch 'main' into text-area-word-limit

francinelucca commented on packages/react/src/components/TextArea/TextArea.tsx:
```diff
@@ -155,32 +161,122 @@ const TextArea = React.forwardRef((props: TextAreaProps, forwardRef) => {
+          evt.preventDefault();
+        }
+      }
+    },
```
```suggestion
    onKeyDown: (evt) => {
      if (!other.disabled && enableCounter && counterMode === 'word') {
        const key = evt.which;
        if (maxCount && textCount >= maxCount && key === 32) {
          evt.preventDefault();
        }
      }
    },
```

francinelucca commented on packages/react/src/components/TextArea/__tests__/TextArea-test.js:
```diff
@@ -309,4 +310,146 @@ describe('TextArea', () => {
+          maxCount={5}
+          counterMode="word"
+          // defaultValue="one two three four five"
+        />
```
```suggestion
      rerender(
        <TextArea
          id="input-1"
          labelText="TextArea label"
          enableCounter={true}
          maxCount={5}
          counterMode="word"
        />
```

francinelucca commented on packages/react/src/components/TextArea/__tests__/TextArea-test.js:
```diff
@@ -309,4 +310,146 @@ describe('TextArea', () => {
+
+      clipboardEvent.clipboardData = {
+        getData: () => 'one two three four five six seven eight nine ten',
+      };
```
is this necessary? doesn't seem like it's being used

matejoslav replied:
let me check this, I know I had one working event and one random one

matejoslav replied:
yep you are right its not used

francinelucca reviewed (changes requested):
the enableCounter bug seems to be fixed 🎉. This is almost there, just noticed one detail:
In the video below, I'd expect the counter to correctly show up the number of words/characters when switching between modes


https://github.com/carbon-design-system/carbon/assets/40550942/07b0cabd-e151-494e-97af-bc8a3156f7d0


Also added some cleanup suggestions :)

matejoslav committed:
Merge remote-tracking branch 'upstream/main' into text-area-word-limit

matejoslav committed:
chore(TextArea): remove console log

matejoslav committed:
chore: Update packages/react/src/components/TextArea/__tests__/TextArea-test.js

Co-authored-by: Francine Lucca <40550942+francinelucca@users.noreply.github.com>

matejoslav committed:
chore: remove redundant event definition in test

matejoslav committed:
Merge branch 'text-area-word-limit' of github.com:matejoslav/carbon into text-area-word-limit

matejoslav commented on packages/react/src/components/TextArea/TextArea.tsx:
```diff
@@ -159,32 +165,122 @@ const TextArea = React.forwardRef((props: TextAreaProps, forwardRef) => {
-    );
-  }, [value, defaultValue]);
+    setTextCount(getInitialTextCount(defaultValue, value, counterMode));
+  }, [value, defaultValue, counterMode]);
```
@francinelucca hmm I would expect this to handle the recounting of characters

matejoslav replied:
Ok the issue is that if value or default value is undefined the change of mode reruns the getInitial count and returns 0, I also discovered that everytime we type the initial count function is reexecuted, that is because there is a re-render of the component (for some reason) and useState reruns the initial value, got a fix I think.

matejoslav committed:
Merge remote-tracking branch 'upstream/main' into text-area-word-limit

matejoslav committed:
fix(TextArea): fix count not updating when switching mode

matejoslav committed:
fix(TextArea): fix onPaste function

matejoslav committed:
fix(TextArea): fix lint errors and tests logic

matejoslav committed:
Merge remote-tracking branch 'upstream/main' into text-area-word-limit

matejoslav committed:
fix(contributors): revert list to what is on main

matejoslav committed:
fix: add contributors

matejoslav committed:
Merge branch 'main' into text-area-word-limit

francinelucca reviewed (approved):
Storybook issue seems to be fixed now, LGTM!

matejoslav committed:
Merge branch 'main' into text-area-word-limit

alisonjoseph reviewed (approved):
LGTM!
